library/items/item-group/src/main/java/org/quiltmc/qsl/itemgroup/api/QuiltItemGroupBuilder.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.itemgroup.api;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.quiltmc.qsl.itemgroup.impl.ItemGroupExtensions;
+
+import net.minecraft.client.gui.screen.ingame.CreativeInventoryScreen;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemGroup;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.collection.DefaultedList;
+
+/**
+ * A builder class that helps with creating and adding {@link ItemGroup}s to the {@link CreativeInventoryScreen}.
+ * <p>
+ * Potential uses:
+ * <p>
+ * Creating an {@link ItemGroup} and then supplying the item icon:
+ * <pre>{@code
+ * public class MyMod implements ModInitializer {
+ *   public static final QuiltItemGroup MY_ITEM_GROUP =
+ *   	new QuiltItemGroupBuilder(
+ *   		new Identifier("my_mod:item_group"))
+ *   	.build();
+ *
+ *   public static Item MY_ITEM;
+ *
+ *   @Override
+ *   public void onInitialize() {
+ *     MY_ITEM = new Item(new QuiltItemSettings().group(MY_ITEM_GROUP));
+ *     MY_ITEM_GROUP.icon(MY_ITEM);
+ *   }
+ * }
+ * }</pre>
+ * <p>
+ * Creating an {@link ItemGroup} with the icon as a supplier:
+ * <pre>{@code
+ * public class MyMod implements ModInitializer {
+ *   public static final QuiltItemGroup MY_ITEM_GROUP =
+ *   	new QuiltItemGroupBuilder(
+ *   		new Identifier("my_mod:item_group"))
+ *   	.icon(() -> new ItemStack(MyMod.MY_ITEM))
+ *   	.build();
+ *
+ *   public static Item MY_ITEM;
+ *
+ *   @Override
+ *   public void onInitialize() {
+ *     MY_ITEM = new Item(new QuiltItemSettings().group(MY_ITEM_GROUP));
+ *   }
+ * }
+ * }</pre>
+ */
+public final class QuiltItemGroupBuilder {
+	private final Identifier identifier;
+	private Supplier<ItemStack> stackSupplier = () -> ItemStack.EMPTY;
+	private Consumer<List<ItemStack>> stacksForDisplay;
+
+	/**
+	 * Create a new {@link QuiltItemGroupBuilder}. Using the constructor allows for the use of the {@link QuiltItemGroupBuilder#icon(Supplier)} and {@link QuiltItemGroupBuilder#appendItems(Consumer)} methods.
+	 * Manually setting the icon with {@link QuiltItemGroup#setIcon(Item)} is possible after calling {@link QuiltItemGroupBuilder#build()}
+	 *
+	 * @param identifier the {@link Identifier} will become the name of the {@link ItemGroup} and will be used for the translation key
+	 */
+	public QuiltItemGroupBuilder(Identifier identifier) {
```

Comment:
This constructor should probably be private given the class is final and we have factory methods